zerver/management/commands/soft_activate_deactivate_users.py

```diff
@@ -63,18 +65,23 @@ def handle(self, *args, **options):
             logger.info('Soft Reactivated %d user(s)' % (len(users_activated)))
         elif deactivate:
             if user_emails:
-                users_to_deactivate = list(UserProfile.objects.filter(
-                    realm=realm,
-                    email__in=user_emails))
+                users_to_deactivate = UserProfile.objects.filter(email__in=user_emails)
+                if options['realm_id']:
+                    users_to_deactivate = users_to_deactivate.filter(realm=realm)
+                users_to_deactivate = list(users_to_deactivate)
                 if len(users_to_deactivate) != len(user_emails):
                     user_emails_found = [user.email for user in users_to_deactivate]
                     for user in user_emails:
                         if user not in user_emails_found:
                             raise Exception('User with email %s was not found. Check if the email is correct.' % (user))
                 print('Soft deactivating forcefully...')
             else:
-                users_to_deactivate = get_users_for_soft_deactivation(realm, int(options['inactive_for']))
-
+                users_to_deactivate = get_users_for_soft_deactivation(int(options['inactive_for']))
+                if options['realm_id']:
+                    users_to_deactivate = [
+                        user for user in users_to_deactivate
+                        if user.realm == realm
+                    ]
```

Comment:
This algorithm will be pretty slow if we ask for a single realm on a large server, since we're filtering in Python, rather in the database.  

What if we instead pass into `get_users_for_soft_deactivation` a `filter_kwargs` argument, that's either `{}` or `dict(user_profile__realm=realm)`, and then add `**filter_kwargs` to the arguments in the `UserActivity.objects.filter()` line in get_users_for_soft_deactivation.